### PR TITLE
Automatically set disk I/O mode to the optimal value (native/threads) 

### LINKED ISSUE #4077: Automatically decide which disk I/O mode should be used
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind enhancement

**What happened**:
KubeVirt supports setting a disk I/O mode for disk devices. This configuration can have great impact on
disk performance but using the wrong configuration can cause negative impact on the performance.

Though `io=native` provides a better performance in many cases (with preallocated disks), it can block the VM
when writing to not fully allocated (sparse) disk. On the other hand, `io=threads` can have negative impact on performance
when we try to use more threads that actual CPU cores. Though advanced users should be able to configure this as they
need, we should have at least some sort of a heuristic to decide on what is the best configuration for each VMI.

A similar approach was taken by the OpenStack team (also includes more details about the problem itself):
https://specs.openstack.org/openstack/nova-specs/specs/mitaka/implemented/libvirt-aio-mode.html

**What you expected to happen**:
We should have a setting on disk devices, something like `io=auto`. When this setting it set, KubeVirt will try
to decide what is the best disk I/O mode for the given disk. This can be done by checking the type of the storage backend
and maybe the number of available cores on the machine on which the VMI will be scheduled on.

### PR DESCRIPTION
**Pull request by ezrasilvera**
**What this PR does / why we need it**:
KubeVirt supports setting a disk I/O mode for disk devices.  There are currently two possible modes 
1.  io=native:  Using kernel async-io. 
1. io=threads: Using user-space threads to achieve (semi) async-I/O

Several investigations showed that on most cases the `native` option results in better I/O performance regardless of the workload and I/O pattern (i.e., requests/sec, size of requests, etc..). 
The only exception was in the case of file based disk with sparse file image. 

This means we can set an optimal disk I/O for the user within KubeVirt independently on the user future usage. 
We will set this mode automatically only in case the user didn't explicitly set that mode himself (we still allow the user to set it up directly)
The logic we take (to make sure we are on the safe side) is:  For block devices, and pre-allocated image disks - use `native`, for everything else don't set the mode explicitly. 

We don't set explicitly `thread` mode for sparse files because there might be use cases in which `native` will out-perform `threads` even for sparse files.

[1] https://web.archive.org/web/20180815010524/http://turlucode.com/qemu-disk-io-performance-comparison-native-or-threads-windows-10-version/
[2] http://docshare02.docshare.tips/files/17877/178771133.pdf
[3] https://www.slideshare.net/pradeepkumarsuvce/qemu-disk-io-which-performs-better-native-or-threads

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4077

**Special notes for your reviewer**:
Not to be confused with `iothreads` which can be orthogonal to this mode
**Release note**:

```release-note
"NONE"
```

### CODE REVIEW

**Comment by rmohr** [comment]
/cc @danielBelenky 

Daniel, could you have a look? I think you investigated exactly this in the past.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 1]
```diff
@@ -194,6 +196,51 @@ func SetDriverCacheMode(disk *Disk) error {
...
 }
 
+func isPreAllocated(path string) bool {
+	preAlloc := false
+	diskInf, err := GetImageInfo(path)
+	if err != nil {
+		return false
+	}
+	// ActualSize can be a little larger then VirtualSize for qcow2
+	if diskInf.VirtualSize < diskInf.ActualSize {
```
Comment (danielBelenky) [inline pkg/virt-launcher/virtwrap/api/converter.go:206]: Maybe for simplicity you could just do:
```
return diskInf.VirtualSize < diskInf.ActualSize
```

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 2]
```diff
@@ -194,6 +196,51 @@ func SetDriverCacheMode(disk *Disk) error {
...
+		preAlloc = true
+	}
+	return preAlloc
+}
+
+// Set optimal io mode automatically
+func SetOptimalIOMode(disk *Disk) error {
+	var path string
+
+	// If the user explicitly set the io mmode do nothing
```
Comment (danielBelenky) [inline pkg/virt-launcher/virtwrap/api/converter.go:216]: You have a typo here: s/mmode/mode

**Review by danielBelenky — CHANGES_REQUESTED** [review]
@ezrasilvera could you please add tests?

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 1 above)
Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/api/converter.go]: Done

File: pkg/virt-launcher/virtwrap/api/converter.go (hunk 2 above)
Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/api/converter.go]: Done

**Comment by ezrasilvera** [comment]
> @ezrasilvera could you please add tests?

In progress - working on this

**Comment by ezrasilvera** [comment]
@danielBelenky I added functional test to test the new IO mode.

**Comment by rmohr** [comment]
/approve

very nice.

**Pull request merged**